File: DOne.js
function str2Arr(str) {
    var result = [];
    for (let i = 0; i < str.length; i++) {
        result.push(str.charCodeAt(i))
    }
    return result;
}
function arr2str(arr) {
    var result = "";
    arr.forEach(charcode => {
        result+=String.fromCharCode(charcode);
    });
    return result;
}
var alphabetBaseMain = str2Arr("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
const DOne = {
    encrypt: function (key, plain) {
        if (typeof plain != 'object')
            throw new Error("Plain type must be a Buffer")
        var result = [];
        var alphabetBase1 = DOne_CreateAlphabetBase(plain);
        alphabetBaseMain = alphabetBase1.alphabet;
        var rest = alphabetBase1.rest;
        for (let i = 0; i < str2Arr(key).length; i++) {
            if (![...alphabetBaseMain].includes(str2Arr(key)[i]))
                throw new Error("Key contain byte that are not in alphabetBase")
        }
        var formatedkey = DOne_FormatKey(key);
        var alphabetList = [];
        for (let i = 0; i < formatedkey.length; i++)
            alphabetList.push(DOne_CreateAlphabet(formatedkey[i]));

        for (let i = 0, o = 0; i < plain.length; i++, o++) {
            var at = alphabetBaseMain.indexOf(plain[i]);
            if (i >= 32)
                o = 0;
            result.push(alphabetList[o][at]);
        }
        return btoa(arr2str(result)) + (rest == "" ? "" : ".") + rest;
    },
    decrypt: function (key, encrypted_b64) {
        var rest = "";
        var encrypted = null;
        if (encrypted_b64.split('.').length == 2) {
            rest = encrypted_b64.split('.')[1];
            encrypted = str2Arr(atob(encrypted_b64.split('.')[0]));
        } else {
            encrypted = str2Arr(atob(encrypted_b64));
        }
        var result = [];
        alphabetBaseMain = DOne_CreateAlphabetBaseFromRest(rest).alphabet;
        for (let i = 0; i < str2Arr(key).length; i++) {
            if (![...alphabetBaseMain].includes(str2Arr(key)[i]))
                throw new Error("Key contain byte that are not in alphabetBase")
        }
        var formatedkey = DOne_FormatKey(key);
        var alphabetList = [];
        for (let i = 0; i < formatedkey.length; i++)
            alphabetList.push(DOne_CreateAlphabet(formatedkey[i]));

        for (let i = 0, o = 0; i < encrypted.length; i++, o++) {
            if (i >= 32)
                o = 0;
            var at = alphabetList[o].indexOf(encrypted[i]);
            result.push(alphabetBaseMain[at]);
        }
        return arr2str(result);
    }
};

function DOne_FormatKey(key) {
    var result2 = [];
    for (let i = 0; result2.length < 32; i++) {
        for (let keyI = 0; keyI < key.length; keyI++) {
            result2.push(str2Arr(key)[keyI]);
        }
        result2 = result2.reverse();
    }
    return result2.slice(0, 32);
}
function DOne_CreateAlphabetBase(plain) {
    var rest = "";
    for (let i = 0; i < plain.length; i++) {
        if (!alphabetBaseMain.includes(plain[i])) {
            rest += String.fromCharCode(plain[i]);
            alphabetBaseMain.push(plain[i]);
        }
    }
    return {
        alphabet: alphabetBaseMain,
        rest: btoa(rest)
    };
}
function DOne_CreateAlphabetBaseFromRest(rest) {
    for (let i = 0; i < str2Arr(atob(rest)).length; i++) {
        if (!alphabetBaseMain.includes(str2Arr(atob(rest))[i])) {
            alphabetBaseMain = alphabetBaseMain.concat(str2Arr(atob(rest))[i]);
        }
    }
    return {
        alphabet: alphabetBaseMain,
        rest: ""
    };
}
function DOne_CreateAlphabet(keybyte) {
    var result = [...alphabetBaseMain];
    var result2 = result.splice(alphabetBaseMain.indexOf(keybyte), result.length - 1);
    // console.log(Uint16Array.from(result.splice(alphabetBaseMain.indexOf(keybyte), result.length-1)).toString());
    return result2.concat(result)
}

File: DOne.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace D_1
{
    class AlphabetBase
    {
        public byte[] alphabet = { };
        public string rest = "";
        public AlphabetBase(byte[] alphabet, string rest)
        {
            this.alphabet = alphabet;
            this.rest = rest;
        }
    }
    class DOne
    {
        private static byte[] alphabetBase = Encoding.UTF8.GetBytes("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
        private static string restBase = "";
        private static AlphabetBase CreateAlphabetBase(byte[] data)
        {
            alphabetBase = Encoding.UTF8.GetBytes("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
            restBase = "";
            List<byte> result = alphabetBase.ToList<byte>();
            List<byte> rest = new List<byte>();
            foreach (byte byteofdata in data)
            {
                if (!result.Contains<byte>(byteofdata))
                {
                    result.Add(byteofdata);
                    rest.Add(byteofdata);
                }
            }
            return new AlphabetBase(result.ToArray(), Convert.ToBase64String(rest.ToArray()));
        }
        private static AlphabetBase CreateAlphabetBaseFromRest(string rest)
        {
            alphabetBase = Encoding.UTF8.GetBytes("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
            restBase = "";
            List<byte> result = alphabetBase.ToList<byte>();
            foreach (byte byteofdata in Convert.FromBase64String(rest))
            {
                if (!result.Contains<byte>(byteofdata))
                {
                    result.Add(byteofdata);
                }
            }
            return new AlphabetBase(result.ToArray(), "");
        }
        public static string Encrypt(byte[] key, byte[] plain)
        {

            List<byte> result = new List<byte>();

            AlphabetBase alphabetBase1 = CreateAlphabetBase(plain);

            alphabetBase = alphabetBase1.alphabet;
            restBase = alphabetBase1.rest;
            foreach (byte keybyte in key)
                if (!alphabetBase.Contains<byte>(keybyte))
                    throw new Exception("Key contain byte that are not in alphabetBase");

            byte[] formatedkey = keyFormating(key);
            List<byte[]> alphabetList = new List<byte[]>();
            foreach (byte keybyte in formatedkey)
                alphabetList.Add(CreateAlphabet(keybyte));


            for (int i = 0, o = 0; i < plain.Length; i++, o++)
            {
                int at = alphabetBase.ToList<byte>().IndexOf(plain[i]);
                if (i >= 32)
                    o = 0;
                result.Add(alphabetList[o][at]);
            }

            return Convert.ToBase64String(result.ToArray())+(restBase == "" ? "" : ".")+restBase;
        }
        public static byte[] Decrypt(byte[] key, string encrypted_b64)
        {
            string rest = "";
            byte[] encrypted = { };
            if(encrypted_b64.Split('.').Length == 2)
            {
                rest = encrypted_b64.Split('.')[1];
                encrypted = Convert.FromBase64String(encrypted_b64.Split('.')[0]);
            } else
            {
                encrypted = Convert.FromBase64String(encrypted_b64);
            }

            List<byte> result = new List<byte>();

            alphabetBase = CreateAlphabetBaseFromRest(rest).alphabet;
            foreach (byte keybyte in key)
                if (!alphabetBase.Contains<byte>(keybyte))
                    throw new Exception("Key contain byte that are not in alphabetBase");

            byte[] formatedkey = keyFormating(key);
            List<byte[]> alphabetList = new List<byte[]>();
            foreach (byte keybyte in formatedkey)
                alphabetList.Add(CreateAlphabet(keybyte));

            for (int i = 0, o = 0; i < encrypted.Length; i++, o++)
            {
                if (i >= 32)
                    o = 0;
                int at = alphabetList[o].ToList<byte>().IndexOf(encrypted[i]);
                result.Add(alphabetBase[at]);
            }
            return result.ToArray();
        }

        private static byte[] CreateAlphabet(byte key)
        {
            return ShiftString(alphabetBase, alphabetBase.ToList<byte>().IndexOf(key));
        }
        private static byte[] keyFormating(byte[] key)
        {
            List<byte> result = key.ToList<byte>();
            for (int i = 0; result.Count < 32; i++)
            {
                foreach (byte keybyte in key)
                {
                    result.Add(keybyte);
                }
                result = ShiftString(result.ToArray(), i).ToList<byte>();
            }
            return Subsequence(result.ToArray(), 0, 32);
        }
        private static byte[] ShiftString(byte[] t, int length)
        {
            List<byte> result = Subsequence(t, length, t.Length - length).ToList<byte>();
            foreach (byte result2byte in Subsequence(t, 0, length).ToList<byte>())
                result.Add(result2byte);
            return result.ToArray();
        }
        private static byte[] Subsequence(byte[] arr, int startIndex, int length)
        {
            return arr.Skip(startIndex).Take(length).ToArray();
        }
    }
}
